### ISSUE TITLE
forlab_io.f90 Fortran runtime error: Attempt to DEALLOCATE unallocated 'self'

### ISSUE THREAD

**Issue by Beliavsky** [fortran-fans/forlab#17]
When I run `fpm build` and `fpm test`, I get the following error:

 + build\gfortran_debug\test\io_bin.exe 
 call randn(X):
  0.6852       0.4069       0.1247E-01   -2.413        2.139    
At line 718 of file .\.\src\forlab_io.f90
Fortran runtime error: Attempt to DEALLOCATE unallocated 'self'

Thanks for the project.

**Comment by Beliavsky** [comment]
The subroutine where the error occurs is

```
    subroutine close (self)

        class(file), intent(out) :: self

        deallocate(self%filename)
        close(self%unit)

    end subroutine close
```

Since `self` is `intent(out)`, I think its components start out unallocated, and you cannot `deallocate` something that is not `allocated`. When I remove the `deallocate` statement there is no error.

**Comment by zoziha** [comment]
Yes, It shall be `class(file), intent(inout) :: self`.

Thank you for your use. I have recently been in contact with fortran-lang's projects stdlib and fftpack. I tried to improve the forlab project before, but it seems that I am not experienced enough and don’t have the idea of engineering. I have studied in fortran-lang for a period of time. Yesterday I just wanted to refactor forlab to make it perfect.
I plan to contribute some more meaningful routines to stdlib. Forlab will use stdlib as a dependency and specify it in the document to determine the survival range of forlab.
